src/ra_utility.h

```diff
@@ -180,85 +120,175 @@ tstrtoul(_In_z_ LPCTSTR _String,
+
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _CONSTANT_FN&
+operator&=(_Inout_ Enum& a, _In_ const Enum b) noexcept { return (a = a & b); }
+
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _NODISCARD _CONSTANT_FN
+operator|(_In_ const Enum a, _In_ const Enum b) noexcept { return(itoe<Enum>(etoi(a) | etoi(b))); }
+
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _CONSTANT_FN&
+operator|=(_Inout_ Enum& a, _In_ const Enum b) noexcept { return (a = a | b); }
+
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _NODISCARD _CONSTANT_FN
+operator^(_In_ const Enum a, _In_ const Enum b) noexcept { return(itoe<Enum>(etoi(a) ^ etoi(b))); }
+
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _CONSTANT_FN&
+operator^=(_Inout_ Enum& a, _In_ const Enum b) noexcept { return (a = a ^ b); }
+
+// The underlying_type must be signed or you could risk an arithmetic overflow
+template<
+    typename Enum,
+    class = std::enable_if_t<std::is_enum_v<Enum> and std::is_signed_v<std::underlying_type_t<Enum>>>
+> _NODISCARD _CONSTANT_FN
+operator~(_In_ const Enum& a) noexcept { return(itoe<Enum>(~etoi(a))); }
+
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _NODISCARD _CONSTANT_FN
+operator<<(_In_ const Enum a, _In_ const std::underlying_type_t<Enum> b) noexcept
 {
-    return (!(a < etoi(b)));
+    // too many risks of using signed numbers with bit shift, so we will make them unsigned no matter what
+    using underlying_type = std::underlying_type_t<Enum>;
+    if constexpr (std::is_signed_v<underlying_type>)
+        return (itoe<Enum>(to_unsigned(etoi(a)) << to_unsigned(b)));
+    else
+        return (itoe<Enum>((etoi(a)) << b));
 }
 
-template<typename Enum, typename = std::enable_if_t<std::is_enum_v<Enum>>> _NODISCARD _CONSTANT_VAR
-operator>=(_In_ const Enum a, _In_ const decltype(etoi(std::declval<Enum&>())) b) noexcept
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _CONSTANT_FN&
+operator<<=(_Inout_ Enum& a, _In_ const std::underlying_type_t<Enum> b) noexcept { return (a = a << b); }
+
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _NODISCARD _CONSTANT_FN
+operator>>(_In_ const Enum a, _In_ const std::underlying_type_t<Enum> b) noexcept
 {
-    return (!(etoi(a) < b));
+    // too many risks of using signed numbers with bit shift, so we will make them unsigned no matter what
+    using underlying_type = std::underlying_type_t<Enum>;
+    if constexpr (std::is_signed_v<underlying_type>)
+        return (itoe<Enum>(to_unsigned(etoi(a)) >> to_unsigned(b)));
+    else
+        return (itoe<Enum>((etoi(a)) >> b));
 }
 
-} // namespace rel_ops
+template<typename Enum, class = std::enable_if_t<std::is_enum_v<Enum>>> _CONSTANT_FN&
+operator>>=(_Inout_ Enum& a, _In_ const std::underlying_type_t<Enum> b) noexcept { return (a = a >> b); }
```

Comment:
These bitwise operators will work for any enum class, constraints were put to filter out invalid input.